ethers-providers/src/provider.rs

```diff
@@ -1161,6 +1161,278 @@ impl<P: JsonRpcClient> Middleware for Provider<P> {
+};
+
+impl<P: JsonRpcClient> Provider<P> {
+    pub fn call_builder<'a>(&'a self, tx: &'a TypedTransaction) -> Caller<'a, P> {
+        Caller::Build(CallBuilder::new(self, tx))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct CallInput<'a> {
+    tx: &'a TypedTransaction,
+    block: Option<BlockId>,
+    spoof: Option<&'a spoof::State>,
+}
+impl<'a> CallInput<'a> {
+    pub fn new(tx: &'a TypedTransaction) -> Self {
+        Self { tx, block: None, spoof: None }
+    }
+}
+
+impl<'a> Serialize for CallInput<'a> {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        let len = 2 + self.spoof.is_some() as usize;
+
+        let mut tup = serializer.serialize_tuple(len)?;
+        tup.serialize_element(self.tx)?;
+
+        let block = self.block.unwrap_or_else(|| BlockNumber::Latest.into());
+        tup.serialize_element(&block)?;
+
+        if let Some(spoof) = self.spoof {
+            tup.serialize_element(spoof)?;
+        }
+        tup.end()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct CallBuilder<'a, P> {
+    provider: &'a Provider<P>,
+    input: CallInput<'a>,
+}
+
+impl<'a, P> CallBuilder<'a, P> {
+    fn new(provider: &'a Provider<P>, tx: &'a TypedTransaction) -> Self {
+        Self { provider, input: CallInput::new(tx) }
+    }
+}
+impl<'a, P: JsonRpcClient> CallBuilder<'a, P> {
+    fn execute(&self) -> impl Future<Output = Result<Bytes, ProviderError>> + 'a {
+        self.provider.request("eth_call", utils::serialize(&self.input))
+    }
+}
+
+pub enum Caller<'a, P> {
+    Build(CallBuilder<'a, P>),
+    Wait(PinBoxFut<'a, Bytes>),
+}
+
+impl<'a, P> Caller<'a, P> {
+    pub fn map<F>(self, f: F) -> Self
+    where
+        F: FnOnce(&mut CallBuilder<'a, P>),
+    {
+        match self {
+            Self::Build(mut call) => {
+                f(&mut call);
+                Self::Build(call)
+            }
+            wait => wait,
+        }
+    }
+    pub fn block(self, id: BlockId) -> Self {
+        self.map(|mut call| call.input.block = Some(id))
+    }
+    pub fn spoof(self, state: &'a spoof::State) -> Self {
+        self.map(|mut call| call.input.spoof = Some(state))
+    }
+    pub fn unwrap_builder(self) -> CallBuilder<'a, P> {
+        match self {
+            Self::Build(b) => b,
+            _ => panic!("Caller::unwrap_builder on a Wait value"),
+        }
+    }
+}
+
+impl<'a, P: JsonRpcClient> Future for Caller<'a, P> {
+    type Output = Result<Bytes, ProviderError>;
+
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context) -> Poll<Self::Output> {
+        let pin = self.get_mut();
+        match pin {
+            Caller::Build(ref call) => {
+                let fut = Box::pin(call.execute());
+                *pin = Caller::Wait(fut);
+                ctx.waker().wake_by_ref();
+                Poll::Pending
+            }
+            Caller::Wait(ref mut fut) => fut.as_mut().poll(ctx),
+        }
```

Comment:
this can be a loop instead of transitioning via a wake?

Reply:
Agreed, loop is much better here.